BenchmarkClient & WorkerImpl

oschaaf opened the pull request:
This PR contains the benchmark client and threader worker

Signed-off-by: Otto van der Schaaf <oschaaf@we-amp.com>

oschaaf commented on nighthawk/source/client/benchmark_client_impl.cc:
```diff
@@ -0,0 +1,230 @@
+    return false;
+  }
+
+  auto stream_decoder = new StreamDecoder(
```
@htuch re: "is it possible that the allocating entity can also do cleanup via a smartptr, or do we need to self-destruct"

I'm not sure, as I understand it `StreamDecoder` relies on Envoy to perform the right callbacks, which will at some point drive it into its terminal state (and `StreamDecoder` is responsible for correctly interpreting these callbacks and determine when it has served its purpose). 

I wonder, wouldn't it be nice if somehow we could optionally transfer or share ownership of the decoder to/with Envoy in the call to `newStream` below? Then we could choose here: we transfer ownership (no pooling), or we share ownership and re-use it  (e.g. recycle to a decoder pool) once Envoy is done with it.

oschaaf replied:
Resolving this as per out-of-band discussion

htuch commented on nighthawk/source/client/benchmark_client_impl.cc:
```diff
@@ -0,0 +1,230 @@
+        store_->createScope("client."), dispatcher_, generator_, *store_, api_,
+        *ssl_context_manager_);
+
+    // TODO(oschaaf): We perform some bootstrapping ourselves here, to avoid an assert during the
```
What do you mean by bootstrapping?

oschaaf replied:
Updated this comment about bootstrapping

htuch commented on nighthawk/source/client/benchmark_client_impl.cc:
```diff
@@ -0,0 +1,230 @@
+    // TODO(oschaaf): We perform some bootstrapping ourselves here, to avoid an assert during the
+    // integration test because of a runtime/tls conflict with the integration test server, when
+    // the message gets validated. Ideally we'd just re-use Tls::Upstream::createTransportFactory()
+
```
Nit: superfluous space.

htuch commented on nighthawk/source/client/benchmark_client_impl.h:
```diff
@@ -0,0 +1,101 @@
+  const std::unique_ptr<Uri> uri_;
+  Envoy::Network::Address::InstanceConstSharedPtr target_address_;
+  std::chrono::seconds timeout_;
+  uint64_t connection_limit_;
```
Tiny nit: I find it easier to verify we didn't skip initializing any of these if you do something like `uint64_t stream_reset_count_{};` to initialize here rather than constructor. For `const`, any non-trivial initializer that depends on args or non-POD, it's fine to do it in constructor.

htuch commented on nighthawk/source/common/worker_impl.h:
```diff
@@ -0,0 +1,39 @@
+  Envoy::ThreadLocal::Instance& tls_;
+  Envoy::Stats::StorePtr store_;
+  std::unique_ptr<Envoy::Runtime::Loader> runtime_;
+  std::unique_ptr<Envoy::Runtime::RandomGenerator> generator_;
```
Should these just be owned objects? Also, Envoy has a singleton runtime now, should we just match that model?

oschaaf replied:
yes, done, now using the singleton runtime.

oschaaf commented on nighthawk/source/client/benchmark_client_impl.cc:
```diff
@@ -0,0 +1,230 @@
+              Envoy::Network::Utility::getAddressWithPort(*address_list.front(), uri_->port());
+          dns_resolved = true;
+        } else {
+          ENVOY_LOG(critical, "Could not resolve host [{}]", uri_->host_without_port());
```
Yes this is tested https://github.com/envoyproxy/envoy-perf/pull/54/files#diff-547dca47cd8b31225f4d73460ac4af1eR280

oschaaf commented on nighthawk/source/client/benchmark_client_impl.cc:
```diff
@@ -0,0 +1,230 @@
+      requests_completed_(0), requests_initiated_(0), measure_latencies_(false) {
+  ASSERT(uri_->isValid());
+
+  connect_statistic_->setId("benchmark_http_client.queue_to_connect");
```
I think it makes sense to take a look at just using Envoy's statistics functionality, and figure out what it takes to switch. One thing I can think of is that we would need to add support for HdrHistogram to Envoy.

htuch commented on nighthawk/include/nighthawk/client/benchmark_client.h:
```diff
@@ -0,0 +1,76 @@
+   * named and value it gets passed.
+   * @return std::string containing zero or more lines containing "name:value\n".
+   */
+  virtual std::string countersToString(CounterFilter filter) const PURE;
```
@oschaaf did we resolve to fix this in this PR or file an issue and do this later?

htuch commented on nighthawk/include/nighthawk/client/benchmark_client.h:
```diff
@@ -0,0 +1,75 @@
+  /**
+   * Determines if latency measurement is on.
+   *
+   * @return true if latency measurement is on.
```
Nit: you can condense this entire comment to:
```
/**
 * @return bool indicating if latency measurement is enabled.
 */
```

htuch commented on nighthawk/include/nighthawk/common/worker.h:
```diff
@@ -0,0 +1,36 @@
+  /**
+   * Perform the actual work on the associated thread initiated by start().
+   */
+  virtual void work() PURE;
```
Why is this in the interface? Isn't it only a `WorkerImpl` detail?

oschaaf replied:
Pinged you on slack to discuss

htuch commented on nighthawk/source/client/benchmark_client_impl.cc:
```diff
@@ -0,0 +1,226 @@
+
+#include "absl/strings/str_split.h"
+
+#include "ares.h"
```
How come we need Ares directly here (vs. via Envoy's DNS libs)?

htuch commented on nighthawk/source/client/benchmark_client_impl.cc:
```diff
@@ -0,0 +1,226 @@
+      transport_socket.set_name(
+          Envoy::Extensions::TransportSockets::TransportSocketNames::get().Tls);
+      Envoy::MessageUtil::jsonConvert(cluster_config.tls_context(),
+                                      *transport_socket.mutable_config());
```
Nit: the more modern way to do this in Envoy is with typed config (`Any`), so that you don't need to round-trip with JSON.

htuch commented on nighthawk/source/client/benchmark_client_impl.cc:
```diff
@@ -0,0 +1,226 @@
+
+StatisticPtrMap BenchmarkClientHttpImpl::statistics() const {
+  StatisticPtrMap statistics;
+  statistics[connect_statistic_->id()] = connect_statistic_.get();
```
Just a placeholder comment re: our discussion on filing an issue on better integration with Envoy stats.

oschaaf replied:
Added to the track-for-future issue

htuch commented on nighthawk/source/client/benchmark_client_impl.cc:
```diff
@@ -0,0 +1,226 @@
+    ASSERT(headers.Status());
+    const int64_t status = Envoy::Http::Utility::getResponseStatus(headers);
+    if (status >= 400 && status <= 599) {
+      // TODO(oschaaf): Figure out why this isn't incremented for us.
```
Happy to do some interactive debug on Slack on this one.

oschaaf replied:
After looking at the code, it looks like this needs to be changed, the stat that is used is incremented upon actual protocol violations, not when the protocol indicates an error.

oschaaf replied:
Added NH's own set of statistics to track counts for 1xx,2xx,..,5xx, resolving this.

htuch commented on nighthawk/source/client/benchmark_client_impl.h:
```diff
@@ -0,0 +1,103 @@
+                          StatisticPtr&& response_statistic, const std::string& uri, bool use_h2);
+
+  // TODO(oschaaf): can probably get rid of these.
+  uint64_t stream_reset_count() { return stream_reset_count_; }
```
Nit: this can be a `const` method`

htuch commented on nighthawk/test/benchmark_http_client_test.cc:
```diff
@@ -0,0 +1,289 @@
+
+namespace Nighthawk {
+
+std::string lorem_ipsum_config;
```
Nit: this is technically a static non-PoD, which can be problematic due to the static initializaiton fiasco in C++. Also, prefer to avoid globals. Can this be inlined into `SetUpTestCase`?

htuch commented on nighthawk/test/benchmark_http_client_test.cc:
```diff
@@ -0,0 +1,289 @@
+  static void SetUpTestCase() {
+    Envoy::Filesystem::InstanceImpl filesystem;
+
+    Envoy::TestEnvironment::setEnvVar("TEST_TMPDIR", Envoy::TestEnvironment::temporaryDirectory(),
```
How come we need to do this? Doesn't Bazel set this?

oschaaf replied:
No, it's not added for us by Bazel, but I removed this here, and added it to the CI script instead, I think that is what Envoy does.

htuch commented on nighthawk/test/benchmark_http_client_test.cc:
```diff
@@ -0,0 +1,289 @@
+  }
+
+  void SetUp() override {
+    ares_library_init(ARES_LIB_INIT_ALL);
```
Consider factoring out this boiler plate in case other tests need it.

oschaaf replied:
Added this to the tracking issue for now

htuch commented on nighthawk/test/benchmark_http_client_test.cc:
```diff
@@ -0,0 +1,289 @@
+    test_server_.reset();
+    fake_upstreams_.clear();
+    tls_.shutdownGlobalThreading();
+    ares_library_cleanup();
```
Ditto.

htuch commented on nighthawk/test/benchmark_http_client_test.cc:
```diff
@@ -0,0 +1,289 @@
+    client_ = std::make_unique<Client::BenchmarkClientHttpImpl>(
+        api_, *dispatcher_, std::make_unique<Envoy::Stats::IsolatedStoreImpl>(),
+        std::make_unique<StreamingStatistic>(), std::make_unique<StreamingStatistic>(),
+        fmt::format("{}://127.0.0.1:{}{}", use_https ? "https" : "http", port, uriPath), use_h2);
```
Nit: this will break in IPv6-only environments such as Google. This will become an issue during the Google import. See how this is handled safely with `TEST_P` and IpVersions in Envoy's `test/integration` tests.

oschaaf replied:
This is fixed, I think, but I'm looking for a way to ensure/check we used `ipv4` / `ipv6` in test expectations.

htuch commented on nighthawk/test/benchmark_http_client_test.cc:
```diff
@@ -0,0 +1,289 @@
+  Envoy::Api::Impl api_;
+  Envoy::Event::DispatcherPtr dispatcher_;
+  Envoy::Runtime::RandomGeneratorImpl generator_;
+  ::testing::NiceMock<Envoy::ThreadLocal::MockInstance> tls_;
```
Nit: probably worth a `using ::testing::NiceMock` at the top of file.

oschaaf replied:
Sure, I'd love to, I thought that `using` should be avoided according to the style guide, but I'm not sure where I picked that up. I propose we do that in a separate PR and to enforce consistency accross the board. Added to the track-for-future list.

htuch commented on nighthawk/test/benchmark_http_client_test.cc:
```diff
@@ -0,0 +1,289 @@
+TEST_P(BenchmarkClientTest, BasicTestHttpsH1) {
+  testBasicFunctionality("/lorem-ipsum-status-200", 1, 1, true, false, 10);
+  EXPECT_EQ("client.ssl.ciphers.ECDHE-RSA-AES128-GCM-SHA256:1\n\
+client.ssl.curves.X25519:1\n\
```
Yeah, I reckon if you get back a proto here, it would be super easy to write individual EXPECTs or do a proto compare.

oschaaf replied:
I switched this to a std::map, I'm planning to consolidate the translation to protoc into a single layer. SG?

htuch commented on nighthawk/test/certs.sh:
```diff
@@ -0,0 +1,60 @@
+#!/bin/bash
```
Can you reuse the certs that already exist in https://github.com/envoyproxy/envoy/tree/master/test/config/integration/certs?

oschaaf replied:
Yes, indeed. I symlinked the certs we need, and after tweaking the environment variables, I was able to drop the script that generates them and use those instead.

htuch commented on nighthawk/test/test_data/benchmark_http_client_test_envoy.yaml:
```diff
@@ -0,0 +1,89 @@
+  access_log_path: /dev/null
+  address:
+    socket_address:
+      address: 127.0.0.1
```
Yeah, I reckon the IPv6 issue will also arise here.

oschaaf replied:
I switched this to `{{ ip_loopback_address }}`, I think that is the the right thing to do here?

oschaaf commented on nighthawk/source/client/benchmark_client_impl.cc:
```diff
@@ -0,0 +1,230 @@
+    if (use_h2_) {
+      common_tls_context->add_alpn_protocols("h2");
+    }
+    common_tls_context->add_alpn_protocols("http/1.1");
```
Yes, good catch, that was actually the way this worked before switching over to the new `ssl.h` we have now.

htuch commented on ci/do_ci.sh:
```diff
@@ -36,6 +36,10 @@ if [ -n "$CIRCLECI" ]; then
         echo 1
     fi
+    export TEST_TMPDIR=/build/tmp
+    export TEST_CERT_DIR="$TEST_TMPDIR"
```
Is `TEST_CERT_DIR` used anywhere?

htuch commented on nighthawk/include/nighthawk/client/benchmark_client.h:
```diff
@@ -0,0 +1,73 @@
+  /**
+   * Gets a map of tracked counter values, keyed by name.
+   * @param filter function that returns true iff a counter should be included in the map,
+   * based on the named and value it gets passed. The default filter returns all counters.
```
FWIW, Envoy also does stats filtering with StatsMatcher.

htuch replied:
https://github.com/envoyproxy/envoy/blob/4a5f85809cef13be47873483789a5a7bb64ded7c/api/envoy/config/metrics/v2/stats.proto#L76

htuch replied:
Actually, maybe I misunderstand. Why do we need to filter here? Could we just return the raw stats object and use that?

oschaaf replied:
We use the filter to get non-0 valued stats. We can use that to keep the output small, but also in tests to ensure that we check all relevant counters (when we want to).

htuch commented on nighthawk/source/client/benchmark_client_impl.cc:
```diff
@@ -0,0 +1,243 @@
+  auto dns_resolver = dispatcher_.createDnsResolver({});
+  std::string to_resolve = uri_->host_without_port();
+
+  if (to_resolve.size() > 0 && to_resolve[0] == '[') {
```
What is this for? Please add comment in source as well as review.

oschaaf replied:
This was checking to see if the hostname represented an ipv6 address; this code changed a bit and now avoids forwarding ip adresses to dns.

htuch commented on nighthawk/source/client/benchmark_client_impl.cc:
```diff
@@ -0,0 +1,243 @@
+                                                 Envoy::Stats::StorePtr&& store,
+                                                 StatisticPtr&& connect_statistic,
+                                                 StatisticPtr&& response_statistic,
+                                                 const std::string& uri, bool use_h2)
```
Tiny nit: generally prefer `absl::string_view` over `const std::string&` in new code, since it's strictly more flexibly and compatible with the C++17 brace new world.

oschaaf replied:
Yeah, added a tracking issue to replace the remaining `const std::string&` instances with `absl::string_view` in one go; resolving this for now.

htuch commented on nighthawk/source/client/benchmark_client_impl.cc:
```diff
@@ -0,0 +1,243 @@
+}
+
+void BenchmarkClientHttpImpl::onPoolFailure(Envoy::Http::ConnectionPool::PoolFailureReason reason) {
+  ASSERT(reason == Envoy::Http::ConnectionPool::PoolFailureReason::ConnectionFailure);
```
Is this more of a `NOT_IMPLEMENTED_UNLESS` or an actual invariant?

oschaaf replied:
We have a bug when this asserts, in the current closed-loop mode we avoid pool overflows as we throttle ourselves. When open-loop testing lands, this needs to be revisited.

htuch commented on nighthawk/source/client/benchmark_client_impl.cc:
```diff
@@ -0,0 +1,243 @@
+    } else if (status > 499 && status <= 599) {
+      benchmark_client_stats_.http_5xx_.inc();
+    } else {
+      benchmark_client_stats_.http_xxx_.inc();
```
Envoy also tracks this, but I think it might be in HCM, which you won't have.

htuch commented on nighthawk/source/client/output.proto:
```diff
@@ -26,4 +24,16 @@ message Statistic {
+message Result {
+  string name = 1;
+  repeated Statistic statistics = 2;
+  repeated Counter counters = 3;
```
Envoy has stats dumping protos, see https://github.com/envoyproxy/envoy/blob/master/api/envoy/admin/v2alpha/metrics.proto, should we reference these instead?

oschaaf replied:
Yes. I looked into that but I'm having trouble with the proto import. Adding this to the list to unblock here.

htuch replied:
Fair enough; would like to see this in the final NH but fine to leave for now.

htuch commented on nighthawk/source/common/utility.cc:
```diff
@@ -34,21 +34,35 @@ Uri::Uri(const std::string& uri) : scheme_("http") {
+  size_t colon_index = std::string::npos;
+  bool in_ipv6_address = false;
+
+  for (size_t i = 0; i < host_and_port_.size(); i++) {
```
Don't any of the existing Envoy address parsing utils help here? I'm surprised we're the first to be deconstructing an IPv6 address..

htuch replied:
If we need this, I suggest splitting out into a separate utility and adding tests..

oschaaf replied:
Well, there's `Utility::parseInternetAddressAndPort()` (which we now use in the client later on). But it expects a port, which we do not always have (the code is working to figure that out here). Splitted this out and added tests; also added some limited tests for hostname validation

htuch replied:
Hmm, this answers my earlier comment. Can't we just do a quick scan from RHS, check if there is a `:`, and if not, add a dummy port?

htuch commented on nighthawk/test/benchmark_http_client_test.cc:
```diff
@@ -0,0 +1,336 @@
+
+    copyFileToWorkingDir("nighthawk/test/test_data/lorem_ipsum.txt", "lorem_ipsum.txt");
+    copyFileToWorkingDir("nighthawk/test/test_data/certs/cacert.pem", "cacert.pem");
+    copyFileToWorkingDir("nighthawk/test/test_data/certs/servercert.pem", "servercert.pem");
```
Why can't we reference these directly in the Envoy tree where they already exist via test data? That way, if we need to bump them, for example when they expire or we want to change some baseline assumptions about server certs, it's less churn.

oschaaf replied:
got rid of these, let me know if this looks good now

htuch commented on nighthawk/test/benchmark_http_client_test.cc:
```diff
@@ -0,0 +1,336 @@
+  }
+
+  static void SetUpTestCase() {
+    ASSERT_NE("", Envoy::TestEnvironment::getCheckedEnvVar("TEST_TMPDIR"));
```
I think you can assume this from Bazel, since it sets this variable, https://docs.bazel.build/versions/master/test-encyclopedia.html

htuch commented on nighthawk/test/benchmark_http_client_test.cc:
```diff
@@ -0,0 +1,336 @@
+      port = use_https ? getTestServerHostAndSslPort() : getTestServerHostAndPort();
+    }
+
+    std::string address =
```
Nit: `const`

htuch commented on nighthawk/test/test_data/benchmark_http_client_test_envoy.yaml:
```diff
@@ -0,0 +1,89 @@
+admin:
```
Did you consider using https://github.com/envoyproxy/envoy/blob/master/test/config/utility.h to build this bootstrap? We try to generally avoid hardcoded bootstrap YAML in tests, but if you need gymnastics to make `ConfigHelper` work, this is probably OK.

oschaaf replied:
Tbh, I had not seen that, it looks useful. I rewrote some of the bootstrap which allowed a reduction of the amount of yaml needed.. Shall we add improving this further to the track-for-future issue?

htuch replied:
Yep, sounds like a plane.

htuch commented on nighthawk/include/nighthawk/client/benchmark_client.h:
```diff
@@ -0,0 +1,73 @@
+   *
+   * @param measure_latencies true iff latencies should be measured.
+   */
+  virtual void setMeasureLatencies(bool measure_latencies) PURE;
```
I wonder if this is exposing an internal implementation detail; why not always measure latencies?

oschaaf replied:
This gets used later, we currently do not measure during the warmup phase, and enable latency when we transition out the it

htuch replied:
Ack, ignore this comment then. Thanks for the explanation.

htuch commented on nighthawk/include/nighthawk/client/benchmark_client.h:
```diff
@@ -0,0 +1,73 @@
+   * @param caller_completion_callback The callback the client must call back upon completion of a
+   * successfully started request.
+   *
+   * @return true if the request could be started.
```
Nit: prefer `@return bool true if .. otherwise ...`

htuch commented on nighthawk/include/nighthawk/client/benchmark_client.h:
```diff
@@ -0,0 +1,73 @@
+
+  /**
+   * Initialize will be called on the worker thread after it has started.
+   * @param runtime to be used during initialization.
```
Nit: document `@return`. What would a fail here do?

oschaaf replied:
This no longer has a return value

htuch commented on nighthawk/source/client/benchmark_client_impl.cc:
```diff
@@ -0,0 +1,247 @@
+    target_address_ = Envoy::Network::Utility::parseInternetAddressAndPort(uri_->host_and_port());
+    return true;
+  } catch (EnvoyException) {
+  }
```
Maybe comment `// We get here when we have a hostname to resolve instead.` in the `catch` braces.

htuch commented on nighthawk/source/common/utility.cc:
```diff
@@ -29,26 +29,73 @@ uint32_t determineCpuCoresWithAffinity() {
+        last_seen_dot = true;
+        continue;
+      } else {
+        last_seen_dot = false;
```
I reckon we probably don't need to do any of this validation; it's the user's responsibility to provide something valid, I think we only care about the `:` separator. Otherwise, we get in the business of doing what other libraries (e.g. Ares) are already doing. This then simplifies quite a bit.

htuch commented on nighthawk/source/common/utility.cc:
```diff
@@ -29,26 +29,73 @@ uint32_t determineCpuCoresWithAffinity() {
+  for (size_t i = 0; i < authority.size(); i++) {
+    char c = authority[i];
+    if (i == 0 && c == '[') {
+      in_ipv6_address = true;
```
I'm probably super dense here, so apologies if we've covered this before, but what's the reason we can't use `Utility::parseInternetAddressAndPort` again?

htuch commented on nighthawk/source/common/utility.cc:
```diff
@@ -29,26 +29,73 @@ uint32_t determineCpuCoresWithAffinity() {
+  try {
+    colon_index = findPortSeparatorInAuthority(host_and_port_);
+  } catch (InvalidHostException ex) {
+    host_error_ = true;
```
Why not throw an exception in the constructor? This is the idiomatic C++ way to fail a constructor.

htuch commented on nighthawk/test/benchmark_http_client_test.cc:
```diff
@@ -0,0 +1,410 @@
+
+  void doSetupBenchmarkClient(const std::string uriPath, bool use_https, bool use_h2) {
+    const std::string address =
+        GetParam() == Envoy::Network::Address::IpVersion::v4 ? "127.0.0.1" : "[::1]";
```
See Envoy's `getLoopbackAddressString` util.

htuch commented on nighthawk/test/benchmark_http_client_test.cc:
```diff
@@ -0,0 +1,410 @@
+    client_->setConnectionLimit(connection_limit);
+    EXPECT_TRUE(client_->initialize(runtime_));
+
+    uint64_t amount = amount_of_request;
```
Nit: `const`

htuch commented on nighthawk/test/test_data/benchmark_http_client_test_envoy.yaml:
```diff
@@ -0,0 +1,49 @@
+                  direct_response:
+                    status: 200
+                    body:
+                      filename: nighthawk/test/test_data/lorem_ipsum.txt
```
Interesting that this works. I think we probably should do a runfiles based substitution here though, since we will need this when we import to Google (the monorepo path will have some additional prefix that we need a hook to substitute).

htuch commented on nighthawk/test/benchmark_http_client_test.cc:
```diff
@@ -0,0 +1,410 @@
+    }
+
+    EXPECT_EQ(max_pending, inflight_response_count);
+
```
Nit: prefer not to have this much superfluous whitepace. Google style prefers compactness.

oschaaf commented:
@htuch this is ready for another pass. I moved all address resolution (parsing, dns) into the Uri class, and simplified it. This fixed another todo: the address resolution in the `nighthawk` branch now happens on the main thread and the resolved uri is passed on to the workers (instead of doing resolution on each worker thread). Also improved the tests for the name resolution process (wrt ipv4/ipv6).

htuch commented on nighthawk/source/common/utility.cc:
```diff
@@ -29,28 +29,97 @@ uint32_t determineCpuCoresWithAffinity() {
   absl::string_view host, path;
   Envoy::Http::Utility::extractHostPathFromUri(uri, host, path);
+
+  if (host.size() == 0) {
```
Nit: `host.empty()`

htuch commented on nighthawk/source/common/utility.cc:
```diff
@@ -29,28 +29,97 @@ uint32_t determineCpuCoresWithAffinity() {
 } // namespace PlatformUtils
 
-Uri::Uri(const std::string& uri) : scheme_("http") {
+size_t Uri::findPortSeparator(absl::string_view hostname) {
```
+1 on this reorg to `Uri`, looks much cleaner now!

htuch commented on nighthawk/source/common/utility.cc:
```diff
@@ -29,28 +29,97 @@ uint32_t determineCpuCoresWithAffinity() {
+            performDnsLookup(dispatcher, dns_lookup_family);
+
+  // Ensure that we figured out a fitting match for the requested dns lookup family.
+  ok = ok && !((dns_lookup_family == Envoy::Network::DnsLookupFamily::V6Only &&
```
Is it possible to trigger these conditions, i.e. if we ask for IPv6 only at DNS resolution, might we actually get back an IPv4?

oschaaf replied:
The test case that attempts to resolve `[::1]:80` while requesting IPv4-only triggers it. But that doesn't go through DNS, and is handled by `Envoy::Network::Utility::parseInternetAddressAndPort()`

htuch commented on nighthawk/test/worker_test.cc:
```diff
@@ -0,0 +1,61 @@
+
+#include "nighthawk/test/mocks.h"
+
+#include "nighthawk/source/common/worker_impl.h"
```
We really need something like Envoy's check/fix_format scripts to enforce header ordering. Can you add this to the backlog if it doesn't already exist there?

htuch commented on nighthawk/test/utility_test.cc:
```diff
@@ -41,17 +46,90 @@ TEST_F(UtilityTest, ExplicitPort) {
+  EXPECT_EQ(80, u2.port());
+}
+
+TEST_F(UtilityTest, findPortSeparator) {
```
Nit: `FindPortSeparator`

htuch commented on nighthawk/test/utility_test.cc:
```diff
@@ -41,17 +46,90 @@ TEST_F(UtilityTest, ExplicitPort) {
+                         testing::ValuesIn(Envoy::TestEnvironment::getIpVersionsForTest()),
+                         Envoy::TestUtility::ipTestParamsToString);
+
+TEST_P(UtilityAddressResolutionTest, addressResolution) {
```
Nit: can you fix up the capitalization on these other tests as well?

htuch commented:
@oschaaf P.S. can you lower the coverage threshold for now so that we can get the coverage HTML and verify the new code is fully tested?

htuch commented on nighthawk/source/common/utility.cc:
```diff
@@ -29,7 +29,7 @@ uint32_t determineCpuCoresWithAffinity() {
 } // namespace PlatformUtils
 
-size_t Uri::findPortSeparator(absl::string_view hostname) {
+size_t Uri::FindPortSeparator(absl::string_view hostname) {
```
I meant change the name in the test, not the implementation :P

htuch commented:
Somehow `run_coverage.sh` is conflicting, can you take a look?

oschaaf commented:
/retest

htuch commented on nighthawk/source/common/utility.cc:
```diff
@@ -29,28 +30,105 @@ uint32_t determineCpuCoresWithAffinity() {
 } // namespace PlatformUtils
 
-Uri::Uri(const std::string& uri) : scheme_("http") {
+Uri Uri::Parse(absl::string_view uri) {
```
FWIW, in a completely independent exercise, I was looking at Chromium's URL handling code, and it has all of this plus more. We're looking at how we can incorporate this Envoy-side long-term, so I feel that at some point, we can just switch to that and not have to maintain our own. Might be worth noting somewhere.
